### DISCUSSION TITLE
Project includes do not handle soft symlinks on linux, instead copy entire linked directory

### DISCUSSION THREAD

**Discussion by levihb** [dotnet/msbuild#6641] [Q&A]
If I have something like the following in my project file:

```csproj
<None Include="fixtures\**" CopyToOutputDirectory="PreserveNewest" LinkBase="fixtures\" />
```

And then inside of fixtures I have a symlinked folder I created using `ln -s [source] [link]`. But the problem is when I build the project dotnet removes the symlink and instead just copies everything over. This doesn't work as when running in debug I need the folder to be elsewhere on the system, both because of access and because it's very large.

Is there a workaround for this?

I put linux in the title, although I have not tested it on any other operating system.

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one area label.

**Comment by rainersigwald** [comment]
No, it doesn't look like `Copy` has a `CopySymlinksAsSymlinks` option. That would be difficult to implement, as the inputs to `Copy` are a list of _files_, which are discovered by the `**` pattern you're specifying, so we don't even have an entry for your symlink--we just thought of it as a directory to traverse.

I think your best bet is to exclude the symlink directory from the pattern and explicitly set up the symlink using `<Exec Command="ln -s ..." />` in a custom target.

**Comment by levihb** [comment]
@rainersigwald Thanks for the response, sorry I didn't get back to it quick enough in the original issue. I appreciate the workaround, but this still seems like the wrong way to handle things to me?

Surely there shouldn't need to even be a `CopySymlinksAsSymlinks`, shouldn't it just be the default behaviour to copy a symlink as a symlink, as happens everywhere else? E.g:

![2021-07-01-15:24:39](https://user-images.githubusercontent.com/42031255/124140671-95461800-da80-11eb-8979-f74ad04c3171.png)

As you can see `cp` will copy the link as a link. msbuild should do the same. In fact it shouldn't even be getting it as a directory, it should be getting it as a file:

![2021-07-01-15:30:51](https://user-images.githubusercontent.com/42031255/124141464-5b294600-da81-11eb-9ba4-ad59bafdf0f0.png)

As you can see above, in linux hard links should be treated as files unless you explicitly treat them as a directory by adding `/` to the end. `ls` was treating the link as a normal file, even with `ls fixtures/**` it was treating it as just a file in that directory.

I feel as if the msbuild implementation is going completely against the other standards here. It should treat a link as a file unless otherwise specified.
